docs/contract/confidential-smart-contract.md

```diff
@@ -0,0 +1,264 @@
+# Confidential Smart Contract
+
+In this chapter we are going to see how to:
+1. write a smart contract which stores and loads data to and from a
+   confidential store and
+2. instantiate and call the smart contract without revealing the call arguments.
+
+## PublicCell VS ConfidentialCell
+
+In the [hello world](./hello-world.md) example we used
+`oasis_contract_sdk_storage::cell::PublicCell<T>` to access the key-value store
+of that contract instance. In this case the value was stored unencrypted on the
+blockchain associated with the hash of the key we provided in the constructor.
+
+Cipher supports another primitive
+`oasis_contract_sdk_storage::cell::ConfidentialCell<T>` which allows you to
+access the data of the contract instance confidentially assured by a
+hardware-level encryption.
+
+In contrast to non-confidential calls the hash of the key we passed to the
+constructor, the runtime library also computes a nonce each time the data is
+accessed. This nonce ensures the confidentiality so that the accessed key
+appears different each time from the blockchain observer and even from the node
+operator. It is generated from:
```

Comment:
This is a bit confusing. So the ever-changing nonce is for the values, not the keys as the keys need to be located at well-known locations. The keys are constant, but the value encryption uses the described nonce scheme.